src/compiler/mir/ops/traverser.rs

```diff
@@ -54,6 +64,42 @@ impl<'a> ProgramTraverser<'a> {
             traverser.map();
         }
     }
+
+    fn map_type<'p, L, V, F: FunctionBuilder<L, V>, P: ProgramBuilder<'p, L, V, F>>(
+        &self,
+        id: TypeId,
+        ty: &MirTypeDef,
+        xfmr: &mut P,
+    ) {
+        debug!("Traversing type {:?}", id);
+
+        // If this is a type that references other types, make sure those referenced types
+        // are defined before transforming this type.
+        match ty {
+            MirTypeDef::Base(_) => (),
+            MirTypeDef::Array { ty, .. } => self.map_type(*ty, self.mir.get_type(*ty), xfmr),
+            MirTypeDef::RawPointer { target, .. } => {
+                let target_ty = self.mir.get_type(*target);
+                match target_ty {
+                    MirTypeDef::Structure { .. } => (),
+                    MirTypeDef::Array { .. }
+                    | MirTypeDef::Base(..)
+                    | MirTypeDef::RawPointer { .. } => self.map_type(*target, target_ty, xfmr),
+                }
+            }
+            MirTypeDef::Structure { def, .. } => match def {
```

Comment:
This is confusing I think.  Should it be moved into a help function or would creating indirect recursion be even more confusing?